packages/filecoin-api/src/aggregator/buffer-reducing.js

```diff
@@ -0,0 +1,263 @@
+ * @param {import('../types').Queue<AggregateOfferMessage>} props.aggregateOfferQueue
+ * @param {string} props.group
+ */
+export async function handleBufferReducingWithAggregate({
+  aggregateInfo,
+  bufferStore,
+  bufferQueue,
+  aggregateOfferQueue,
+  group,
+}) {
+  // If aggregate has enough space
+  // store buffered pieces that are part of aggregate and queue aggregate
+  // store remaining pieces and queue them to be reduced
+  /** @type {import('./api.js').Buffer} */
+  const aggregateReducedBuffer = {
+    aggregate: aggregateInfo.aggregate.link,
+    pieces: aggregateInfo.addedBufferedPieces,
+    group,
+  }
+  const aggregateBlock = await CBOR.write(aggregateReducedBuffer)
+
+  // Store buffered pieces for aggregate
+  const bufferStoreAggregatePut = await bufferStore.put({
+    buffer: aggregateReducedBuffer,
+    block: aggregateBlock.cid,
+  })
+  if (bufferStoreAggregatePut.error) {
+    return {
+      error: new StoreOperationFailed(bufferStoreAggregatePut.error.message),
+    }
+  }
+
+  // Propagate message for aggregate offer queue
+  const aggregateOfferQueueAdd = await aggregateOfferQueue.add({
+    // TODO task CID drop?
+    task: aggregateInfo.aggregate.link,
+    aggregate: aggregateInfo.aggregate.link,
+    pieces: aggregateBlock.cid,
+    group,
+  })
+  if (aggregateOfferQueueAdd.error) {
+    return {
+      error: new QueueOperationFailed(aggregateOfferQueueAdd.error.message),
+    }
+  }
+
+  // Store remaining buffered pieces to reduce if they exist
+  if (!aggregateInfo.remainingBufferedPieces.length) {
+    return { ok: {} }
+  }
+
+  const remainingReducedBuffer = {
+    pieces: aggregateInfo.remainingBufferedPieces,
+    group: group,
+  }
+  const remainingBlock = await CBOR.write(remainingReducedBuffer)
+
+  // Store remaining buffered pieces
+  const bufferStoreRemainingPut = await bufferStore.put({
+    buffer: remainingReducedBuffer,
+    block: remainingBlock.cid,
+  })
+  if (bufferStoreRemainingPut.error) {
+    return {
+      error: new StoreOperationFailed(bufferStoreRemainingPut.error.message),
+    }
+  }
+
+  // Propagate message for buffer queue
+  const bufferQueueAdd = await bufferQueue.add({
+    pieces: remainingBlock.cid,
+    group: group,
+  })
+  if (bufferQueueAdd.error) {
+    return {
+      error: new QueueOperationFailed(bufferQueueAdd.error.message),
+    }
+  }
+
+  return { ok: {} }
+}
+
+/**
+ * Store given buffer into store and queue it to further reducing.
+ *
+ * @param {object} props
+ * @param {import('./api.js').Buffer} props.buffer
+ * @param {import('../types').Store<Link, BufferRecord>} props.bufferStore
+ * @param {import('../types').Queue<BufferMessage>} props.bufferQueue
+ */
+export async function handleBufferReducingWithoutAggregate({
+  buffer,
+  bufferStore,
+  bufferQueue,
+}) {
+  const block = await CBOR.write(buffer)
+
+  // Store block in buffer store
+  const bufferStorePut = await bufferStore.put({
+    buffer,
```

Comment:
The buffer store should be responsible for storing underneath as it would like to. Being a dynamo or a bucket should not be a concern here at the interface level. Moreover, if we consider the actual implementation as a bucket, if we would provide bytes we would not be able to store by a given key. So, let's just be consistent within all our stores and let the implementer decide how to store the data